apps/admin/backend/src/exports/csv_results.ts

```diff
@@ -21,135 +22,295 @@ import { ScannerBatch } from '../types';
 import { Store } from '../store';
 import { tabulateElectionResults } from '../tabulation/full_results';
 
-function generateHeaders({
+/**
+ * Possible metadata attributes that can be included in the CSV, in the order
+ * the columns will appear in the CSV.
+ */
+const METADATA_ATTRIBUTES = [
+  'precinct',
+  'party',
+  'ballotStyle',
+  'votingMethod',
+  'scanner',
+  'batch',
+] as const;
+
+type MetadataAttribute = typeof METADATA_ATTRIBUTES[number];
+
+const METADATA_ATTRIBUTE_COMPOUND_LABEL: Record<MetadataAttribute, string> = {
+  precinct: 'Precincts',
+  party: 'Parties',
+  ballotStyle: 'Ballot Styles',
+  votingMethod: 'Voting Methods',
+  scanner: 'Scanners',
+  batch: 'Batches',
+};
+
+// `all` is equivalent to having no filter for an attribute
+type Multiplicity = 'single' | 'multi' | 'all';
+
+/**
+ * Describes the metadata structure of the CSV. We distinguish between single
+ * multi attributes because "single" is the common and intuitive use case,
+ * alongside which we want to add additional related metadata. The "multi"
+ * attributes are only to ensure the CSV is self-documenting when there are
+ * filters selecting for multiple values.
+ */
+type MetadataStructure = {
+  [K in MetadataAttribute]: Multiplicity;
+};
+
+/**
+ * Exported for testing purposes only.
+ */
+export function determineMetadataStructure({
+  filter,
   groupBy,
-  isPrimaryElection,
 }: {
+  filter: Tabulation.Filter;
   groupBy: Tabulation.GroupBy;
-  isPrimaryElection: boolean;
+}): MetadataStructure {
+  const filterStructure: MetadataStructure = {
+    precinct: filter.precinctIds
+      ? filter.precinctIds.length > 1
+        ? 'multi'
+        : 'single'
```

Comment:
The case where `filter.precinctIds.length === 0` is irrelevant because in that case there would be no report, because no records match that filter.